clash-lib/src/Clash/Core/Termination.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Clash.Core.Termination
+  ( RecInfo
+  , mkRecInfo
+  , isRecursive
+  , recursiveGroup
+  ) where
+
+import Control.Lens.Fold
+import Data.Graph (SCC(..))
+import qualified Data.Graph as Graph
+import qualified Data.List as List
+
+import Clash.Core.FreeVars
+import Clash.Core.Var
+import Clash.Core.VarEnv
+import Clash.Driver.Types
+
+-- Quick lookup for whether a binding is recursive or non-recursive. If a
+-- binding is non-recursive, we can assume that it terminates and skip
+-- analysing it.
+--
+data RecInfo = RecInfo
+  { recBindings    :: [VarSet]
+    -- ^ Recursive bindings, organized into groups of strongly connected
+    -- components.
```

Comment:
Clash Core only has self-recursive binders, not mutually recursive ones. I believe that happens over here: 

https://github.com/clash-lang/clash-compiler/blob/305427f3e2bb44669ae971c4d0e34da6a7818ed0/clash-ghc/src-ghc/Clash/GHC/GenerateBindings.hs#L154-L197

Reply:
To say that Clash Core only has self-recursive bindings is a bit of a stretch. We ensure that the list of bindings given to `generateHDL` only has self-recursive bindings by turning any recursive group:
```haskell
f = <fBody>
g = <gBody>
h = <hBody>
```
into:
```haskell
f = let g = <gBody>; h = <hBody> in <fBody>
g = let f = <fBody>; h = <hBody> in <gBody>
h = let f = <fBody>; g = <gBody> in <hBody>
```
and then try to take great care that we never introduce mutually recursive global bindings as we perform all the normalization transformations. I would be preferable if we actually dealt with mutually recursive bindings properly; so this is actually a step in the right direction.